Update Marketplace contract to use CEI pattern

shahthepro opened the pull request:
Should close #3413 

The unit tests don't fail. However, I'm yet to test the compiled contract code with the DApp.

### Checklist:

- [ ] Test your work and double-check to confirm that you didn't break anything
- [ ] [Wrap any new text/strings with `fbt`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#wrapping-text) for translation
- [ ] If there are any new/changed strings to translate, [`npm run translate`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#integrating-translations)
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/docs)

shahthepro commented:
@franckc @nick 

Just FYI, The following code calls two method. Both of these methods follow the CEI pattern. But when calling them in sequential order, it is not technically following CEI. Should I rewrite this? Or leave it as it is?

https://github.com/OriginProtocol/origin/blob/ec434d79c4b0064378330eff4b1e6e0d3fdffdf8/packages/contracts/contracts/marketplace/V01_Marketplace.sol#L227-L243

franckc commented:
Nice work @shahthepro - it gives a much more precise understanding of the changes involved to implement CEI.

Regarding your specific question about what to do for the makeOffer method that also handles withdrawing a prior offer. It looks like that would involve rewriting this whole method and duplicating a bunch of code from withdrawOffer and makeOffer. It increases the chance that we introduce a new bug as we rewrite this method... I'm actually wondering if we do need to keep this method around. As far as I can tell, we are currently not using it anywhere in the DApp ? Is anyone aware of a specific use case for which we know we'll need this method in the future ?

nick commented on packages/contracts/contracts/marketplace/V01_Marketplace.sol:
```diff
@@ -310,14 +311,18 @@ contract V01_Marketplace is Ownable {
-        } else {
+        delete offers[listingID][offerID];
+
+        if (msg.sender != offer.buyer) {
```
Any reason we swapped these around?

shahthepro replied:
`listing.deposit += offer.commission;` is an effect (state change). As per CEI, this effect should happen before interactions with other contracts (which `payCommission` and `paySeller` do)

nick commented:
makeOffer with withdraw essentially allows you to update an offer. If both 'sub' methods follow the CEI pattern... does calling them one after another not also follow it? If not, does that mean all our proxy methods for calling multiple methods one after the other are also vulnerable?

nick commented:
Can we try running Slither on the new contract?

franckc commented:
Thanks @nick for your input despite being a sleep deprived new dad! :)
A few things:
- The main goal of following the CEI pattern is to protect against reentrancy attack if a token that has external code execution (such as ERC 223) is used on the marketplace. So if 2 sub-methods M1 and M2 follow CEI, I don't think the caller of those 2 methods is CEI compliant since reentrancy could occur at the end of the call to M1, before M2 gets executed.
 - Agreed we should do review of our proxy code with this vulnerability in mind. Let's do that separately from this change. I'll follow up with the eng team.
 - Do you have any opinion on whether it's worth rewriting the "update offer" method (aka make offer with withdrawal) to be CEI compliant or if we should drop it for now since it is not currently used in the DApp ?

nick commented:
> reentrancy could occur at the end of the call to M1, before M2 gets executed.

Hmm yea perhaps it could occur but I'm not sure if it's actually vulnerable since internal state regarding balances etc are dealt with before the external call is made. Maybe Slither would flag it if it's an issue?

> rewriting the "update offer" method

I think 'update offer' is an important method. Hopefully we don't need to re-write it if we're not vulnerable to the above.

mikeshultz commented:
> So if 2 sub-methods M1 and M2 follow CEI, I don't think the caller of those 2 methods is CEI compliant since reentrancy could occur at the end of the call to M1, before M2 gets executed.

While true, I'm wondering what actual effect this would have other than a revert?  The checks on the offer object (sender and status) would just fail in the `require()` calls since the offer no longer exists, no?

nick commented:
I guess slither doesn't like the change. If we can get slither to pass, maybe we're OK:

```
marketplace $ slither V01_Marketplace.sol 
INFO:Detectors:
V01_Marketplace.refundBuyer(address,ERC20,uint256) (V01_Marketplace.sol#381-390) sends eth to arbitrary user
	Dangerous calls:
	- require(bool,string)(buyer.send(value),ETH refund failed) (V01_Marketplace.sol#383)
V01_Marketplace.paySeller(address,address,ERC20,uint256,uint256) (V01_Marketplace.sol#393-409) sends eth to arbitrary user
	Dangerous calls:
	- require(bool,string)(buyer.send(offerRefund),ETH refund failed) (V01_Marketplace.sol#397)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#functions-that-send-ether-to-arbitrary-destinations
INFO:Detectors:
Reentrancy in V01_Marketplace.makeOffer(uint256,bytes32,uint256,address,uint256,uint256,ERC20,address,uint256) (V01_Marketplace.sol#227-243):
	External calls:
	- withdrawOffer(listingID,_withdrawOfferID,_ipfsHash) (V01_Marketplace.sol#241)
	- makeOffer(listingID,_ipfsHash,_finalizes,_affiliate,_commission,_value,_currency,_arbitrator) (V01_Marketplace.sol#242)
	External calls sending eth:
	- withdrawOffer(listingID,_withdrawOfferID,_ipfsHash) (V01_Marketplace.sol#241)
	State variables written after the call(s):
	- offers (V01_Marketplace.sol#242)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities
```

nick commented:
Alternatively we can just remove the method and not allow atomic updating of offers 🤷‍♂

franckc commented:
Let's make a decision so that this PR can get merged and tested a bit before we send the contract code over for another round of review (we are trying to send the audit company our response by EOD Thu).

Agreed the VM should revert if there is reentrancy in the withdrawOffer method since the offer gets deleted. So upon re-entry, trying to load the offer should cause a revert I believe. I haven't tested it though. We could demonstrate that and let the security audit company know.

Personally I'm fine with removing this method since it is for future hypothetic use and it is the simplest thing to do for now.

@shahthepro I think when we chatted last night you were also leaning towards removing the method ? Feel free to make the call...

shahthepro commented:
@franckc I removed that method and fixed a couple of other potential bugs reported by slither.

Slither still shows a couple of errors

![Screenshot 2019-09-19 at 10 03 49 AM](https://user-images.githubusercontent.com/10547529/65213682-cf6d4200-dac4-11e9-9aa1-3f08a2359885.png)

It shows these two errors since we pass the `address` as parameter. But these two methods are private and I backtraced it to see if there are any way to forward an arbitrary address to these method. AFAICT, It cannot be done. On publicly exposed methods, we set the seller or buyer address to `msg.sender`. So this is safe to ignore.

Here is the complete output of slither. The rest seem like suggestions on casing and changing few `public` scoped functions to `external`. That shouldn't cause us any trouble though
```
INFO:Detectors:
V01_Marketplace.refundBuyer(address,ERC20,uint256) (contracts/marketplace/V01_Marketplace.sol#368-377) sends eth to arbitrary user
	Dangerous calls:
	- require(bool,string)(buyer.send(value),ETH refund failed) (contracts/marketplace/V01_Marketplace.sol#370)
V01_Marketplace.paySeller(address,address,ERC20,uint256,uint256) (contracts/marketplace/V01_Marketplace.sol#380-396) sends eth to arbitrary user
	Dangerous calls:
	- require(bool,string)(buyer.send(offerRefund),ETH refund failed) (contracts/marketplace/V01_Marketplace.sol#384)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#functions-that-send-ether-to-arbitrary-destinations
INFO:Detectors:
Pragma version "^0.4.24" allows old versions (contracts/marketplace/V01_Marketplace.sol#1)
Pragma version "^0.4.24" allows old versions (node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol#1)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity
INFO:Detectors:
Contract 'V01_Marketplace' (contracts/marketplace/V01_Marketplace.sol#19-449) is not in CapWords
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#82) is not in mixedCase
Parameter '_deposit' of _deposit (contracts/marketplace/V01_Marketplace.sol#82) is not in mixedCase
Parameter '_depositManager' of _depositManager (contracts/marketplace/V01_Marketplace.sol#82) is not in mixedCase
Parameter '_seller' of _seller (contracts/marketplace/V01_Marketplace.sol#90) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#91) is not in mixedCase
Parameter '_deposit' of _deposit (contracts/marketplace/V01_Marketplace.sol#92) is not in mixedCase
Parameter '_depositManager' of _depositManager (contracts/marketplace/V01_Marketplace.sol#93) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#132) is not in mixedCase
Parameter '_additionalDeposit' of _additionalDeposit (contracts/marketplace/V01_Marketplace.sol#133) is not in mixedCase
Parameter '_seller' of _seller (contracts/marketplace/V01_Marketplace.sol#139) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#141) is not in mixedCase
Parameter '_additionalDeposit' of _additionalDeposit (contracts/marketplace/V01_Marketplace.sol#142) is not in mixedCase
Parameter '_target' of _target (contracts/marketplace/V01_Marketplace.sol#172) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#172) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#185) is not in mixedCase
Parameter '_finalizes' of _finalizes (contracts/marketplace/V01_Marketplace.sol#186) is not in mixedCase
Parameter '_affiliate' of _affiliate (contracts/marketplace/V01_Marketplace.sol#187) is not in mixedCase
Parameter '_commission' of _commission (contracts/marketplace/V01_Marketplace.sol#188) is not in mixedCase
Parameter '_value' of _value (contracts/marketplace/V01_Marketplace.sol#189) is not in mixedCase
Parameter '_currency' of _currency (contracts/marketplace/V01_Marketplace.sol#190) is not in mixedCase
Parameter '_arbitrator' of _arbitrator (contracts/marketplace/V01_Marketplace.sol#191) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#233) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#251) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#265) is not in mixedCase
Parameter '_value' of _value (contracts/marketplace/V01_Marketplace.sol#265) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#286) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#316) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#333) is not in mixedCase
Parameter '_ruling' of _ruling (contracts/marketplace/V01_Marketplace.sol#334) is not in mixedCase
Parameter '_refund' of _refund (contracts/marketplace/V01_Marketplace.sol#335) is not in mixedCase
Parameter '_refund' of _refund (contracts/marketplace/V01_Marketplace.sol#357) is not in mixedCase
Parameter '_ipfsHash' of _ipfsHash (contracts/marketplace/V01_Marketplace.sol#357) is not in mixedCase
Parameter '_tokenAddr' of _tokenAddr (contracts/marketplace/V01_Marketplace.sol#434) is not in mixedCase
Parameter '_affiliate' of _affiliate (contracts/marketplace/V01_Marketplace.sol#439) is not in mixedCase
Parameter '_affiliate' of _affiliate (contracts/marketplace/V01_Marketplace.sol#445) is not in mixedCase
Parameter '_newOwner' of _newOwner (node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol#51) is not in mixedCase
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#conformance-to-solidity-naming-conventions
INFO:Detectors:
V01_Marketplace.acceptOffer (contracts/marketplace/V01_Marketplace.sol#233-248) uses literals with too many digits:
	- offer.finalizes < 1000000000
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#too-many-digits
INFO:Detectors:
V01_Marketplace.totalListings() (contracts/marketplace/V01_Marketplace.sol#72-74) should be declared external
V01_Marketplace.totalOffers(uint256) (contracts/marketplace/V01_Marketplace.sol#77-79) should be declared external
V01_Marketplace.createListing(bytes32,uint256,address) (contracts/marketplace/V01_Marketplace.sol#82-86) should be declared external
V01_Marketplace.createListingWithSender(address,bytes32,uint256,address) (contracts/marketplace/V01_Marketplace.sol#89-100) should be declared external
V01_Marketplace.updateListing(uint256,bytes32,uint256) (contracts/marketplace/V01_Marketplace.sol#130-136) should be declared external
V01_Marketplace.updateListingWithSender(address,uint256,bytes32,uint256) (contracts/marketplace/V01_Marketplace.sol#138-149) should be declared external
V01_Marketplace.withdrawListing(uint256,address,bytes32) (contracts/marketplace/V01_Marketplace.sol#172-180) should be declared external
V01_Marketplace.makeOffer(uint256,bytes32,uint256,address,uint256,uint256,ERC20,address) (contracts/marketplace/V01_Marketplace.sol#183-230) should be declared external
V01_Marketplace.acceptOffer(uint256,uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#233-248) should be declared external
V01_Marketplace.withdrawOffer(uint256,uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#251-262) should be declared external
V01_Marketplace.addFunds(uint256,uint256,bytes32,uint256) (contracts/marketplace/V01_Marketplace.sol#265-283) should be declared external
V01_Marketplace.finalize(uint256,uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#286-313) should be declared external
V01_Marketplace.dispute(uint256,uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#316-327) should be declared external
V01_Marketplace.executeRuling(uint256,uint256,bytes32,uint256,uint256) (contracts/marketplace/V01_Marketplace.sol#330-354) should be declared external
V01_Marketplace.updateRefund(uint256,uint256,uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#357-365) should be declared external
V01_Marketplace.addData(bytes32) (contracts/marketplace/V01_Marketplace.sol#409-411) should be declared external
V01_Marketplace.addData(uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#414-416) should be declared external
V01_Marketplace.addData(uint256,uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#419-421) should be declared external
V01_Marketplace.sendDeposit(uint256,address,uint256,bytes32) (contracts/marketplace/V01_Marketplace.sol#424-431) should be declared external
V01_Marketplace.addAffiliate(address,bytes32) (contracts/marketplace/V01_Marketplace.sol#439-442) should be declared external
V01_Marketplace.removeAffiliate(address,bytes32) (contracts/marketplace/V01_Marketplace.sol#445-448) should be declared external
Ownable.renounceOwnership() (node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol#42-45) should be declared external
Ownable.transferOwnership(address) (node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol#51-53) should be declared external
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#public-function-that-could-be-declared-as-external
INFO:Slither:contracts/marketplace/V01_Marketplace.sol analyzed (3 contracts), 65 result(s) found
```
